src/lock.rs

```diff
@@ -27,30 +27,25 @@ use crate::{
 
 pub const LOCKFILE: &str = "Proto.lock";
 
-const SHA256_TAG: &str = "sha256";
-
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum DigestAlgorithm {
+    #[serde(rename = "sha256")]
     SHA256,
 }
 
-impl std::fmt::Display for DigestAlgorithm {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let s = match self {
-            Self::SHA256 => SHA256_TAG,
-        };
-        f.write_str(s)
+impl FromStr for DigestAlgorithm {
+    type Err = serde_typename::Error;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        serde_typename::from_str(input)
     }
 }
 
-impl FromStr for DigestAlgorithm {
-    type Err = eyre::Report;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            SHA256_TAG => Ok(Self::SHA256),
-            other => eyre::bail!("invalid digest algorithm: {other}"),
-        }
+impl fmt::Display for DigestAlgorithm {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // this will not fail as long as `DigestAlgorithm` is an enum,
+        let typename = serde_typename::to_str(self).expect("digest algorithm type name");
```

Comment:
Can we map this as a format result error and not expect?

Reply:
We technically can, but [std::fmt::Error](https://doc.rust-lang.org/stable/std/fmt/struct.Error.html) does not let us add any information about what went wrong:

> This type does not support transmission of an error other than that an error occurred. Any extra information must be arranged to be transmitted through some other means.

Realistically, this can never panic, as long as `DigestAlgorithm` is an enum. We can check that in unit tests. Perhaps a semantically better way to write this would be:

```rust
match serde_typename::to_string(self) {
    Ok(value) => value,
    Err(error) => unreachable!("DigestAlgorithm failed to convert to string: {error}"),
}
```

That way we would convey that this is not something that should ever occur.

Reply:
For what it's worth I think it's perfectly reasonable to expect here, it's what expect is intended for. But I've had this discussion with Mara before.

Reply:
I have changed it to use `unreachable!()`, I find that that is more clear anyways. Whenever I see `expect`, it makes me wonder if someone forgot to do error handling here or if this is truly something that cannot happen. So `unreachable` is a lot more clear in this case.